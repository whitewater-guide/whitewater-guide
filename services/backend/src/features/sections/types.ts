import { RawTimestamped } from '@db';
import { PointRaw } from '@features/points';
import {
  Coordinate3d,
  NamedNode,
  SectionEditAction,
  Tag,
} from '@whitewater-guide/commons';

export interface GaugeBindingRaw {
  minimum?: number;
  maximum?: number;
  optimum?: number;
  impossible?: number;
  approximate?: boolean;
}

interface ShapeRaw {
  type: string;
  coordinates: Coordinate3d[];
}

export interface SectionRaw extends NamedNode, RawTimestamped {
  alt_names: string[];

  river_id: string;
  river_name: string; // for embedding
  gauge_id: string | null;
  region_id: string;
  region_name: string; // for embedding

  hidden: boolean;
  demo: boolean;
  premium: boolean;

  season_numeric: number[];
  levels: GaugeBindingRaw | null;
  flows: GaugeBindingRaw | null;
  shape: ShapeRaw;
  distance: number | null;
  drop: number | null;
  duration: number | null;
  difficulty: number;
  difficulty_xtra: string | null;
  rating: number | null;
  description: string | null;
  season: string | null;
  flows_text: string | null;
  put_in: string;
  take_out: string;

  pois: PointRaw[] | null;
  tags: Tag[] | null;

  count: number | null; // window function count
}

export interface SectionsFilter {
  riverId?: string;
  regionId?: string;
  updatedAfter?: Date;
}

export interface SectionsEditLogRaw {
  id?: string; // optional for input
  section_id: string;
  old_section_name: string;
  new_section_name: string;
  river_id: string;
  river_name: string;
  region_id: string;
  region_name: string;
  editor_id: string;
  editor_name: string;
  action: SectionEditAction;
  created_at: Date;
}
