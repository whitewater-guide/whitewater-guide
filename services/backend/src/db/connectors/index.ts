export * from './types';
export * from './connector';
export * from './createConnectors';
export * from './queryBuilder';
