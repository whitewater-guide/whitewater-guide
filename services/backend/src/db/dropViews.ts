import Knex from 'knex';

export async function dropViews(db: Knex, ...views: string[]) {
  for (const view of views) {
    await db.raw(`DROP VIEW IF EXISTS ${view}_view`);
  }
}
