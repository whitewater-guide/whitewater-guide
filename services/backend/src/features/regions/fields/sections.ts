import { Context, ListQuery } from '~/apollo';
import { SectionsFilter } from '~/features/sections';
import { GraphQLFieldResolver } from 'graphql';
import { RegionRaw } from '../types';

interface Vars extends ListQuery {
  filter?: SectionsFilter;
}

const sectionsResolver: GraphQLFieldResolver<RegionRaw, Context, Vars> = async (
  { id },
  { page, filter = {} },
  { dataSources },
  info,
) => {
  const { updatedAfter } = filter;
  return dataSources.sections.getMany(info, {
    page,
    filter: { regionId: id, updatedAfter },
  });
};

export default sectionsResolver;
