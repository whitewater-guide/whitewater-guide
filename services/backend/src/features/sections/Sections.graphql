type GaugeBinding {
  # Below this level the river is not enjoyable to paddle
  minimum: Float
  # The river is still runnable above this level, but difficulty is considerably higher than mentioned here
  maximum: Float
  # The best level for paddlers who consider themselves to be in the same class as the section
  optimum: Float
  # It is insanely dangerous or physically impossible to run river this high
  impossible: Float
  # If true, then the gauge gives very vague idea of actual section flows (e.g. the gauge is on other tributary)
  approximate: Boolean
  # Optional formula to compute section flow from gauge value, e.g '(x + 10) * 2' where x is gauge value
  formula: String
}

input GaugeBindingInput {
  minimum: Float
  maximum: Float
  optimum: Float
  impossible: Float
  approximate: Boolean
  formula: String
}

type Section {
  # Unique section ID
  id: ID!
  # Human-friendly section name (without river name, e.g. just 'Upper')
  name: String!
  # Alternative names, e.g. english transliteration for names that contain umlauts
  altNames: [String!]
  # Guidebook page. Subject to premium access limitations
  # Returns null if requires premium access
  # Returns empty string if no description provided
  description: String
  # Human-friendly description to complement numeric months list
  season: String
  # Array of numers between 0 and 23, corresponding to half-months
  seasonNumeric: [Int]!

  # Section region
  region: Region
  # Section river
  river: River

  # Section gauge
  gauge: Gauge
  # Level bindings bindings
  levels: GaugeBinding
  # Flow gauge bindings
  flows: GaugeBinding
  # Human-friendly description to complement gauge bindings, e.g. 'After heavy rainfall'
  flowsText: String

  # Put-in location
  putIn: Point!
  # Take-out location
  takeOut: Point!
  # Array of [lon, lat, alt] points to roughly match river shape, ordered (inclusively) from put-in to take-out
  shape: [[Float!]]!
  # Approximate section length, in km
  distance: Float
  # Approximate difference between put-in and take-out altitudes, in m
  drop: Float
  # Dictionay value, approximate time required to run this section
  duration: Int
  # Floating point difficulty value, i.e. 4.5 means IV+
  difficulty: Float!
  # Short extra difficulty to be written in brackets, i.e. what's inside brackets for "IV (2xV, X)" difficuly
  difficultyXtra: String
  # Subjective rating from 0 to 5 stars, half stars allowed
  rating: Float
  # Array of tags
  tags: [Tag]!
  # When was this section added to whitewater.guide
  createdAt: Date
  # When was this section last modified
  updatedAt: Date
  # Photos, videos and blogs
  media(page: Page): SectionMediaConnection!
  # Points of interest
  pois: [Point]!

  # Not ready for publication yet
  hidden: Boolean!
  # Is demo section in premium region
  demo: Boolean
}

type SectionMediaConnection {
  nodes: [Media]!
  count: Int!
}

input SectionInput {
  id: ID
  name: String!
  altNames: [String!]
  description: String
  season: String
  seasonNumeric: [Int]

  river: RefInput!
  gauge: RefInput
  # Optional, but required when creating new river
  region: RefInput
  levels: GaugeBindingInput
  flows: GaugeBindingInput
  flowsText: String

  shape: [[Float!]]!
  distance: Float
  drop: Float
  duration: Int
  difficulty: Float!
  difficultyXtra: String
  rating: Float
  tags: [RefInput]!
  pois: [PointInput]!

  hidden: Boolean!
}

input SectionAdminSettings {
  demo: Boolean!
}

type SectionsList {
  nodes: [Section]!
  count: Int!
}

input SectionsFilter {
  riverId: ID
  regionId: ID
  updatedAfter: Date
}

type SectionEditLogEntry {
  id: ID!
  section: Section!
  editor: User!
  action: String!
  diff: JSON
  createdAt: Date!
}

input SectionsEditLogFilter {
  regionId: ID
  editorId: ID
}

type SectionsEditLogList {
  nodes: [SectionEditLogEntry]!
  count: Int!
}

type Query {
  sections(page: Page, filter: SectionsFilter): SectionsList!
  section(id: ID): Section
  sectionsEditLog(
    filter: SectionsEditLogFilter
    page: Page
  ): SectionsEditLogList! @admin
}

type Mutation {
  upsertSection(section: SectionInput!): Section
  removeSection(id: ID!): ID
  administrateSection(id: ID!, settings: SectionAdminSettings!): Section @admin
}
