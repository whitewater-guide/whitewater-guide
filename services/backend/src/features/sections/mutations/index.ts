import administrateSection from './administrateSection';
import bulkInsert from './bulkInsert';
import removeSection from './removeSection';
import upsertSection from './upsertSection';

export default {
  administrateSection,
  bulkInsert,
  removeSection,
  upsertSection,
};
