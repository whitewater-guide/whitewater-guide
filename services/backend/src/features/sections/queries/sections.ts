import { ListQuery, TopLevelResolver } from '@apollo';
import { SectionsFilter } from '../types';

interface Vars extends ListQuery {
  filter?: SectionsFilter;
}

const sections: TopLevelResolver<Vars> = async (
  _,
  { filter, page },
  { dataSources },
  info,
) => {
  const result = await dataSources.sections.getMany(info, { filter, page });
  return result;
};

export default sections;
