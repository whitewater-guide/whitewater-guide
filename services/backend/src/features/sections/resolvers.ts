import { listResolvers } from '@apollo';
import { gaugeBindingResolver, sectionFieldResolvers } from './fields';
import Mutation from './mutations';
import Query from './queries';

export const sectionsResolvers = {
  Section: sectionFieldResolvers,
  GaugeBinding: gaugeBindingResolver,
  SectionsList: listResolvers,
  SectionMediaConnection: listResolvers,
  Query,
  Mutation,
};
