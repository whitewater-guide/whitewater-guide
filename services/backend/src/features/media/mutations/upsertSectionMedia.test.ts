import db, { holdTransaction, rollbackTransaction } from '@db';
import { SectionsEditLogRaw } from '@features/sections';
import {
  fileExistsInBucket,
  MEDIA,
  resetTestMinio,
  TEMP,
  TEMP_BUCKET_DIR,
} from '@minio';
import {
  ADMIN,
  ADMIN_ID,
  EDITOR_GA_EC,
  EDITOR_NO_EC,
  EDITOR_NO_EC_ID,
  TEST_USER,
} from '@seeds/01_users';
import { REGION_NORWAY } from '@seeds/04_regions';
import { RIVER_SJOA } from '@seeds/07_rivers';
import { GALICIA_R1_S1, NORWAY_SJOA_AMOT } from '@seeds/09_sections';
import { PHOTO_1, PHOTO_2 } from '@seeds/11_media';
import {
  anonContext,
  countRows,
  fakeContext,
  noTimestamps,
  noUnstable,
  runQuery,
  UUID_REGEX,
} from '@test';
import {
  ApolloErrorCodes,
  MediaInput,
  MediaKind,
} from '@whitewater-guide/commons';
import { copy } from 'fs-extra';
import * as path from 'path';

let mBefore: number;
let msBefore: number;
let trBefore: number;

beforeAll(async () => {
  [mBefore, msBefore, trBefore] = await countRows(
    true,
    'media',
    'sections_media',
    'media_translations',
  );
});
beforeEach(async () => {
  await holdTransaction();
  await resetTestMinio();
});
afterEach(rollbackTransaction);
afterAll(() => resetTestMinio(true));

const mutation = `
  mutation upsertSectionMedia($sectionId: ID!, $media: MediaInput!){
    upsertSectionMedia(sectionId: $sectionId, media: $media){
      id
      kind
      description
      copyright
      url
      resolution
      weight
      createdAt
      updatedAt
      size
    }
  }
`;

const sectionId = NORWAY_SJOA_AMOT;

const NEW_MEDIA_ID = '717b3058-28f0-11e8-b467-0ed5f89f718b';

const media: MediaInput = {
  id: NEW_MEDIA_ID, // always has some id, pregenerated by mediaForm
  copyright: 'new media copyright',
  description: 'new media description',
  kind: MediaKind.photo,
  url: NEW_MEDIA_ID,
  resolution: [1920, 1080],
  weight: null,
};

describe('resolvers chain', () => {
  it('anon should not pass', async () => {
    const result = await runQuery(
      mutation,
      { sectionId, media },
      anonContext(),
    );
    expect(result).toHaveGraphqlError(ApolloErrorCodes.UNAUTHENTICATED);
  });

  it('user should not pass', async () => {
    const result = await runQuery(
      mutation,
      { sectionId, media },
      fakeContext(TEST_USER),
    );
    expect(result).toHaveGraphqlError(ApolloErrorCodes.FORBIDDEN);
  });

  it('non-owning editor should not pass', async () => {
    const result = await runQuery(
      mutation,
      { sectionId, media },
      fakeContext(EDITOR_GA_EC),
    );
    expect(result).toHaveGraphqlError(ApolloErrorCodes.FORBIDDEN);
  });

  it('admin should pass', async () => {
    const result = await runQuery(
      mutation,
      { sectionId, media },
      fakeContext(ADMIN),
    );
    expect(result).not.toHaveGraphqlError(undefined);
  });

  it('should throw on invalid input', async () => {
    const badMedia = {
      id: 'wtf',
      copyright: 'new media copyright',
      description: 'new media description',
      kind: MediaKind.photo,
      url: 'zzzzz',
      resolution: [800, 600, 900],
      weight: -10,
    };
    const result = await runQuery(
      mutation,
      { sectionId, media: badMedia },
      fakeContext(EDITOR_NO_EC),
    );
    expect(result).toHaveGraphqlValidationError();
  });
});

describe('insert', () => {
  it('should fail on non-existing section id', async () => {
    const result = await runQuery(
      mutation,
      { sectionId: '852421bc-2848-11e8-b467-0ed5f89f718b', media },
      fakeContext(ADMIN),
    );
    expect(result).toHaveGraphqlError(ApolloErrorCodes.BAD_USER_INPUT);
  });

  it('should return result', async () => {
    const result = await runQuery(
      mutation,
      { sectionId, media },
      fakeContext(EDITOR_NO_EC),
    );
    expect(result).not.toHaveGraphqlError();
    expect(noTimestamps(result.data)).toMatchSnapshot();
  });

  it('should add one more media to db', async () => {
    await runQuery(mutation, { sectionId, media }, fakeContext(EDITOR_NO_EC));
    const [m, ms, tr] = await countRows(
      false,
      'media',
      'sections_media',
      'media_translations',
    );
    expect([m - mBefore, ms - msBefore, tr - trBefore]).toMatchObject([
      1,
      1,
      1,
    ]);
  });

  it('should sanitize input', async () => {
    const dirty = { ...media, description: "it's a \\ $1 slash with . ?" };
    const result = await runQuery(
      mutation,
      { sectionId, media: dirty },
      fakeContext(EDITOR_NO_EC),
    );
    expect(result).toHaveProperty(
      'data.upsertSectionMedia.description',
      "it's a \\ $1 slash with . ?",
    );
  });

  it('should set created_by field', async () => {
    await runQuery(mutation, { sectionId, media }, fakeContext(EDITOR_NO_EC));
    const { created_by } = await db(false)
      .table('media')
      .select(['created_by'])
      .where({ id: NEW_MEDIA_ID })
      .first();
    expect(created_by).toBe(EDITOR_NO_EC_ID);
  });

  it('should log this event', async () => {
    await runQuery(mutation, { sectionId, media }, fakeContext(EDITOR_NO_EC));
    const entry: SectionsEditLogRaw = await db(false)
      .table('sections_edit_log')
      .orderBy('created_at', 'desc')
      .select('*')
      .first();
    expect(entry).toMatchObject({
      id: expect.stringMatching(UUID_REGEX),
      section_id: NORWAY_SJOA_AMOT,
      old_section_name: 'Amot',
      new_section_name: 'Amot',
      river_id: RIVER_SJOA,
      river_name: 'Sjoa',
      region_id: REGION_NORWAY,
      region_name: 'Norway',
      editor_id: EDITOR_NO_EC_ID,
      action: 'media_create',
      created_at: expect.any(Date),
    });
  });
});

describe('update', () => {
  const uMedia = { ...media, id: PHOTO_1 };

  it('should fail on wrong section id', async () => {
    const result = await runQuery(
      mutation,
      { sectionId: GALICIA_R1_S1, media: uMedia },
      fakeContext(EDITOR_GA_EC),
    );
    expect(result).toHaveGraphqlError(ApolloErrorCodes.BAD_USER_INPUT);
  });

  it('should return result', async () => {
    const result = await runQuery(
      mutation,
      { sectionId, media: uMedia },
      fakeContext(EDITOR_NO_EC),
    );
    expect(result.errors).toBeUndefined();
    expect(noUnstable(result.data)).toMatchSnapshot();
  });

  it('should not change db counts', async () => {
    await runQuery(
      mutation,
      { sectionId, media: uMedia },
      fakeContext(EDITOR_NO_EC),
    );
    const [m, ms, tr] = await countRows(
      false,
      'media',
      'sections_media',
      'media_translations',
    );
    expect([m - mBefore, ms - msBefore, tr - trBefore]).toMatchObject([
      0,
      0,
      0,
    ]);
  });

  it('should not change media kind', async () => {
    const badMedia = { ...uMedia, kind: 'blog' };
    const result = await runQuery(
      mutation,
      { sectionId, media: badMedia },
      fakeContext(EDITOR_NO_EC),
    );
    expect(result.errors).toBeUndefined();
    expect(result).toHaveProperty('data.upsertSectionMedia.kind', 'photo');
  });

  it('should not modify created_by', async () => {
    await runQuery(
      mutation,
      { sectionId, media: uMedia },
      fakeContext(EDITOR_NO_EC),
    );
    const { created_by } = await db(false)
      .table('media')
      .select(['created_by'])
      .where({ id: uMedia.id })
      .first();
    expect(created_by).toBe(ADMIN_ID);
  });

  it('should log this event', async () => {
    await runQuery(
      mutation,
      { sectionId, media: uMedia },
      fakeContext(EDITOR_NO_EC),
    );
    const entry: SectionsEditLogRaw = await db(false)
      .table('sections_edit_log')
      .orderBy('created_at', 'desc')
      .select('*')
      .first();
    expect(entry).toMatchObject({
      id: expect.stringMatching(UUID_REGEX),
      section_id: NORWAY_SJOA_AMOT,
      old_section_name: 'Amot',
      new_section_name: 'Amot',
      river_id: RIVER_SJOA,
      river_name: 'Sjoa',
      region_id: REGION_NORWAY,
      region_name: 'Norway',
      editor_id: EDITOR_NO_EC_ID,
      action: 'media_update',
      created_at: expect.any(Date),
    });
  });
});

describe('i18n', () => {
  const aMedia = { ...media, id: PHOTO_2 };
  const uMedia = { ...media, id: PHOTO_1 };

  it('should add new translation', async () => {
    await runQuery(
      mutation,
      { sectionId, media: aMedia },
      fakeContext(EDITOR_NO_EC, 'ru'),
    );
    const [m, ms, tr] = await countRows(
      false,
      'media',
      'sections_media',
      'media_translations',
    );
    expect([m - mBefore, ms - msBefore, tr - trBefore]).toMatchObject([
      0,
      0,
      1,
    ]);
  });

  it('should modify common props in other language', async () => {
    await runQuery(
      mutation,
      { sectionId, media: aMedia },
      fakeContext(EDITOR_NO_EC, 'ru'),
    );
    const { resolution } = await db()
      .table('media_view')
      .select('resolution')
      .where({ id: PHOTO_2, language: 'en' })
      .first();
    expect(resolution).toMatchObject([1920, 1080]);
  });

  it('should modify existing translation', async () => {
    await runQuery(
      mutation,
      { sectionId, media: uMedia },
      fakeContext(EDITOR_NO_EC, 'ru'),
    );
    const [m, ms, tr] = await countRows(
      false,
      'media',
      'sections_media',
      'media_translations',
    );
    expect([m - mBefore, ms - msBefore, tr - trBefore]).toMatchObject([
      0,
      0,
      0,
    ]);
    const { description } = await db()
      .table('media_view')
      .select('description')
      .where({ id: PHOTO_1, language: 'ru' })
      .first();
    expect(description).toBe('new media description');
  });
});

describe('files', () => {
  let result: any;

  beforeEach(async () => {
    // Emulate file with the same key as existing avatar uploaded by user
    await copy(
      path.resolve(__dirname, '__tests__/test.jpg'),
      path.resolve(TEMP_BUCKET_DIR, NEW_MEDIA_ID),
    );
    result = await runQuery(
      mutation,
      { sectionId, media },
      fakeContext(EDITOR_NO_EC),
    );
    expect(result.errors).toBeUndefined();
  });

  it('should move files from temp to media', async () => {
    await expect(fileExistsInBucket(TEMP, 'test.jpg')).resolves.toBe(false);
    await expect(fileExistsInBucket(TEMP, NEW_MEDIA_ID)).resolves.toBe(false);
    // expect items bucket to contain new_image.jpg
    await expect(
      fileExistsInBucket(
        MEDIA,
        NEW_MEDIA_ID,
        'a1c4720fa8526d4a8560dd1cb29c0ea7',
      ),
    ).resolves.toBe(true);
  });

  it('should store only filenames', async () => {
    const { url } = await db()
      .table('media')
      .select(['url'])
      .where({ id: NEW_MEDIA_ID })
      .first();
    expect(url).toBe(NEW_MEDIA_ID);
  });

  it('should store media size', async () => {
    const { size } = await db()
      .table('media')
      .select(['size'])
      .where({ id: NEW_MEDIA_ID })
      .first();
    expect(size).toBe(14935);
  });

  it('should return filename as url', async () => {
    expect(result).toHaveProperty('data.upsertSectionMedia.url', NEW_MEDIA_ID);
  });

  it('should return media size', async () => {
    expect(result).toHaveProperty('data.upsertSectionMedia.size', 14935);
  });
});
