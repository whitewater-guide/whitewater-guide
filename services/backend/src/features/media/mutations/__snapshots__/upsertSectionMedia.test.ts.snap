// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insert should return result 1`] = `
Object {
  "upsertSectionMedia": Object {
    "copyright": "new media copyright",
    "description": "new media description",
    "id": "717b3058-28f0-11e8-b467-0ed5f89f718b",
    "kind": "photo",
    "resolution": Array [
      1920,
      1080,
    ],
    "size": 0,
    "url": "717b3058-28f0-11e8-b467-0ed5f89f718b",
    "weight": 0,
  },
}
`;

exports[`update should return result 1`] = `
Object {
  "upsertSectionMedia": Object {
    "copyright": "new media copyright",
    "description": "new media description",
    "kind": "photo",
    "resolution": Array [
      1920,
      1080,
    ],
    "size": 0,
    "url": "717b3058-28f0-11e8-b467-0ed5f89f718b",
    "weight": 0,
  },
}
`;
